File: .workbuddy/AiApiTest.java
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import javax.crypto.Cipher;
import javax.crypto.spec.SecretKeySpec;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.KeyFactory;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.SecureRandom;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;
import java.util.Base64;

/**
 * AI 接口测试脚本 (模拟前端加密流程)
 * 用法:
 *   java AiApiTest captcha            -> 获取验证码 保存图片到 captcha.png 并输出 uuid
 *   java AiApiTest login <code> <uuid> -> 带验证码登录 输出 token
 */
public class AiApiTest {

    static final String RSA_PUBLIC_KEY = "MFwwDQYJKoZIhvcNAQEBBQADSwAwSAJBAKoR8mX0rGKLqzcWmOzbfj64K8ZIgOdHnzkXSOVOZbFu/TJhZ7rFAN+eaGkl3C4buccQd/EjEsj9ir7ijT7h96MCAwEAAQ==";
    static final String RSA_PRIVATE_KEY = "MIIBVAIBADANBgkqhkiG9w0BAQEFAASCAT4wggE6AgEAAkEAqhHyZfSsYourNxaY7Nt+PrgrxkiA50efORdI5U5lsW79MmFnusUA355oaSXcLhu5xxB38SMSyP2KvuKNPuH3owIDAQABAkAfoiLyL+Z4lf4Myxk6xUDgLaWGximj20CUf+5BKKnlrK+Ed8gAkM0HqoTt2UZwA5E2MzS4EI2gjfQhz5X28uqxAiEA3wNFxfrCZlSZHb0gn2zDpWowcSxQAgiCstxGUoOqlW8CIQDDOerGKH5OmCJ4Z21v+F25WaHYPxCFMvwxpcw99EcvDQIgIdhDTIqD2jfYjPTY8Jj3EDGPbH2HHuffvflECt3Ek60CIQCFRlCkHpi7hthhYhovyloRYsM+IS9h/0BzlEAuO0ktMQIgSPT3aFAgJYwKpqRYKlLDVcflZFCKY7u3UP8iWi1Qw0Y=";
    static final String BASE = "http://localhost:8080";
    static final String CLIENT_ID = "e5cd7e4891bf95d1d19206ce24a7b32e";

    public static void main(String[] args) throws Exception {
        HttpClient client = HttpClient.newHttpClient();
        if ("captcha".equals(args[0])) {
            HttpRequest request = HttpRequest.newBuilder().uri(URI.create(BASE + "/auth/code"))
                .header("clientid", CLIENT_ID).GET().build();
            HttpResponse<String> resp = client.send(request, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
            String body = resp.body();
            Files.writeString(Path.of("captcha_raw.json"), body);
            // 提取 uuid 与 img
            String uuid = extract(body, "\"uuid\":\"");
            String img = extract(body, "\"img\":\"");
            Files.write(Path.of("captcha.png"), Base64.getDecoder().decode(img));
            System.out.println("UUID=" + uuid);
            System.out.println("captcha.png saved");
        } else if ("peek".equals(args[0])) {
            // 直连 Redis 读取验证码真实值: java AiApiTest peek <uuid>
            try (Socket socket = new Socket("192.168.150.10", 6379)) {
                OutputStream out = socket.getOutputStream();
                InputStream in = socket.getInputStream();
                send(out, "AUTH", "tomori");
                System.out.println("AUTH -> " + readLine(in));
                send(out, "SELECT", "6");
                System.out.println("SELECT -> " + readLine(in));
                send(out, "GET", "GLOBAL:CAPTCHA_CODES:" + args[1]);
                String status = readLine(in);
                System.out.println("GET status -> " + status);
                if (status != null && status.startsWith("$")) {
                    int len = Integer.parseInt(status.substring(1));
                    if (len >= 0) {
                        byte[] buf = new byte[len];
                        int read = 0;
                        while (read < len) {
                            read += in.read(buf, read, len - read);
                        }
                        in.read(); in.read(); // \r\n
                        System.out.println("VALUE -> " + new String(buf, StandardCharsets.UTF_8));
                    }
                }
            }
        } else if ("login".equals(args[0])) {
            String code = args[1];
            String uuid = args[2];
            String body = "{\"clientId\":\"" + CLIENT_ID + "\",\"grantType\":\"password\",\"tenantId\":\"000000\","
                + "\"username\":\"admin\",\"password\":\"666666\",\"code\":\"" + code + "\",\"uuid\":\"" + uuid + "\"}";
            System.out.println(postEncrypted(client, "/auth/login", body));
        }
    }

    static void send(OutputStream out, String... args) throws Exception {
        StringBuilder sb = new StringBuilder("*" + args.length + "\r\n");
        for (String arg : args) {
            sb.append('$').append(arg.getBytes(StandardCharsets.UTF_8).length).append("\r\n").append(arg).append("\r\n");
        }
        out.write(sb.toString().getBytes(StandardCharsets.UTF_8));
        out.flush();
    }

    static String readLine(InputStream in) throws Exception {
        StringBuilder sb = new StringBuilder();
        int c;
        while ((c = in.read()) != -1) {
            if (c == '\r') {
                in.read();
                break;
            }
            sb.append((char) c);
        }
        return sb.toString();
    }

    static String extract(String json, String key) {
        int start = json.indexOf(key) + key.length();
        int end = json.indexOf("\"", start);
        return json.substring(start, end);
    }

    /** 模拟前端加密: AES/ECB 加密 body, RSA 加密 AES key 放 encrypt-key 头 */
    static String postEncrypted(HttpClient client, String path, String body) throws Exception {
        String chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
        SecureRandom random = new SecureRandom();
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < 32; i++) {
            sb.append(chars.charAt(random.nextInt(chars.length())));
        }
        String aesKey = sb.toString();

        Cipher aes = Cipher.getInstance("AES/ECB/PKCS5Padding");
        aes.init(Cipher.ENCRYPT_MODE, new SecretKeySpec(aesKey.getBytes(StandardCharsets.UTF_8), "AES"));
        String encBody = Base64.getEncoder().encodeToString(aes.doFinal(body.getBytes(StandardCharsets.UTF_8)));

        byte[] keyBytes = Base64.getDecoder().decode(RSA_PUBLIC_KEY);
        PublicKey publicKey = KeyFactory.getInstance("RSA").generatePublic(new X509EncodedKeySpec(keyBytes));
        Cipher rsa = Cipher.getInstance("RSA/ECB/PKCS1Padding");
        rsa.init(Cipher.ENCRYPT_MODE, publicKey);
        String encKey = Base64.getEncoder().encodeToString(
            rsa.doFinal(Base64.getEncoder().encodeToString(aesKey.getBytes(StandardCharsets.UTF_8)).getBytes(StandardCharsets.UTF_8)));

        // 本地自解密验证 (模拟后端 DecryptRequestBodyWrapper)
        PrivateKey privateKey = KeyFactory.getInstance("RSA")
            .generatePrivate(new PKCS8EncodedKeySpec(Base64.getDecoder().decode(RSA_PRIVATE_KEY)));
        Cipher rsaDec = Cipher.getInstance("RSA/ECB/PKCS1Padding");
        rsaDec.init(Cipher.DECRYPT_MODE, privateKey);
        String aesKeyRecovered = new String(Base64.getDecoder().decode(
            new String(rsaDec.doFinal(Base64.getDecoder().decode(encKey)), StandardCharsets.UTF_8)), StandardCharsets.UTF_8);
        Cipher aesDec = Cipher.getInstance("AES/ECB/PKCS5Padding");
        aesDec.init(Cipher.DECRYPT_MODE, new SecretKeySpec(aesKeyRecovered.getBytes(StandardCharsets.UTF_8), "AES"));
        String plainBody = new String(aesDec.doFinal(Base64.getDecoder().decode(encBody)), StandardCharsets.UTF_8);
        System.out.println("[SELF-DECRYPT] " + plainBody);

        HttpRequest request = HttpRequest.newBuilder()
            .uri(URI.create(BASE + path))
            .header("Content-Type", "application/json;charset=utf-8")
            .header("clientid", CLIENT_ID)
            .header("encrypt-key", encKey)
            .POST(HttpRequest.BodyPublishers.ofString(encBody, StandardCharsets.UTF_8))
            .build();
        HttpResponse<String> resp = client.send(request, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
        return "HTTP " + resp.statusCode() + "\n" + resp.body();
    }
}

File: .workbuddy/ai-api-test.js
/**
 * RuoYi-Vue-Plus 后端 AI 功能接口测试脚本 (Node 24+, 零依赖)
 *
 * 用法:
 *   node ai-api-test.js login   # 加密登录获取 token (验证码通过 Redis 直读自动解析)
 *   node ai-api-test.js ai      # 测试 AI 接口: 创建会话/会话列表/历史消息/流式对话
 *   node ai-api-test.js ping    # 检查本地 1234 端口的 OpenAI 兼容模型服务
 *   node ai-api-test.js all     # 依次执行 login -> ping -> ai
 */
const crypto = require('crypto');
const fs = require('fs');
const net = require('net');
const path = require('path');

// ============ 配置 ============
const BASE = 'http://localhost:8080';
// RSA 公钥与后端 api-decrypt privateKey 对应 (见 ruoyi-ui/.env.development)
const RSA_PUBLIC_KEY = 'MFwwDQYJKoZIhvcNAQEBBQADSwAwSAJBAKoR8mX0rGKLqzcWmOzbfj64K8ZIgOdHnzkXSOVOZbFu/TJhZ7rFAN+eaGkl3C4buccQd/EjEsj9ir7ijT7h96MCAwEAAQ==';
const CLIENT_ID = 'e5cd7e4891bf95d1d19206ce24a7b32e';
const USERNAME = 'admin';
const PASSWORD = 'admin123';
// Redis (验证码存储) 见 application-dev.yml
const REDIS = { host: '192.168.150.10', port: 6379, password: 'tomori', db: 6 };
// OpenAI 兼容模型服务 (ai.enabled 时后端调用)
const LLM_BASE = 'http://192.168.100.1:1234';
const LLM_MODEL = 'qwen3.5-2b';
const TOKEN_FILE = path.join(__dirname, 'token.txt');

// ============ 加密 (与前端 utils/crypto.ts + jsencrypt 一致) ============
function generateAesKey() {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  let key = '';
  for (let i = 0; i < 32; i++) key += chars.charAt(Math.floor(Math.random() * chars.length));
  return key;
}

function encryptWithAes(plain, key) {
  // AES/ECB/PKCS7 -> Base64
  const cipher = crypto.createCipheriv('aes-256-ecb', Buffer.from(key, 'utf8'), null);
  return Buffer.concat([cipher.update(plain, 'utf8'), cipher.final()]).toString('base64');
}

function encryptWithRsa(plain) {
  const pem = `-----BEGIN PUBLIC KEY-----\n${RSA_PUBLIC_KEY.match(/.{1,64}/g).join('\n')}\n-----END PUBLIC KEY-----`;
  return crypto.publicEncrypt({ key: pem, padding: crypto.constants.RSA_PKCS1_PADDING }, Buffer.from(plain, 'utf8')).toString('base64');
}

/** 模拟前端加密流程发送 POST 请求 */
async function postEncrypted(urlPath, bodyObj, headers = {}) {
  const aesKey = generateAesKey();
  const encBody = encryptWithAes(JSON.stringify(bodyObj), aesKey);
  const encKey = encryptWithRsa(Buffer.from(aesKey, 'utf8').toString('base64'));
  const resp = await fetch(BASE + urlPath, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json;charset=utf-8',
      clientid: CLIENT_ID,
      'encrypt-key': encKey,
      ...headers
    },
    body: encBody
  });
  return { status: resp.status, data: await resp.json(), headers: resp.headers };
}

// ============ Redis RESP 直读验证码 ============
function redisGetValue(key) {
  return new Promise((resolve, reject) => {
    const socket = net.connect(REDIS.port, REDIS.host);
    let phase = 0;
    let pending = '';
    socket.on('connect', () => socket.write(respCmd('AUTH', REDIS.password)));
    socket.on('data', (chunk) => {
      pending += chunk.toString();
      if (phase === 0 && pending.endsWith('+OK\r\n')) {
        phase = 1; pending = '';
        socket.write(respCmd('SELECT', String(REDIS.db)));
      } else if (phase === 1 && pending.endsWith('+OK\r\n')) {
        phase = 2; pending = '';
        socket.write(respCmd('GET', key));
      } else if (phase === 2 && pending.endsWith('\r\n')) {
        const m = pending.match(/^\$(-?\d+)\r\n([\s\S]*?)\r\n$/);
        socket.destroy();
        if (m) {
          // TypedJsonJacksonCodec 存储字符串带 JSON 引号
          resolve(m[1] === '-1' ? null : m[2].replace(/^"|"$/g, ''));
        } else if (pending.startsWith('$-1')) {
          resolve(null);
        }
      }
    });
    socket.on('error', reject);
    setTimeout(() => { socket.destroy(); reject(new Error('redis timeout')); }, 5000);
  });
}

function respCmd(...args) {
  let s = `*${args.length}\r\n`;
  for (const a of args) s += `$${Buffer.byteLength(a)}\r\n${a}\r\n`;
  return s;
}

// ============ 登录 ============
async function login() {
  console.log('==> [1/3] 获取验证码');
  const codeResp = await fetch(BASE + '/auth/code', { headers: { clientid: CLIENT_ID } });
  const codeData = await codeResp.json();
  if (codeData.code !== 200) throw new Error('验证码接口异常: ' + JSON.stringify(codeData));
  const uuid = codeData.data.uuid;

  console.log('==> [2/3] Redis 直读验证码答案 (uuid=' + uuid + ')');
  const answer = await redisGetValue('GLOBAL:CAPTCHA_CODES:' + uuid);
  if (!answer) throw new Error('验证码已从 Redis 过期');
  console.log('    验证码答案: ' + answer);

  console.log('==> [3/3] 加密登录 (' + USERNAME + ')');
  const { status, data } = await postEncrypted('/auth/login', {
    clientId: CLIENT_ID,
    grantType: 'password',
    tenantId: '000000',
    username: USERNAME,
    password: PASSWORD,
    code: answer,
    uuid
  });
  if (data.code !== 200) throw new Error('登录失败: ' + data.msg);
  const token = data.data.access_token;
  fs.writeFileSync(TOKEN_FILE, token);
  console.log('    登录成功, token 已保存 -> ' + TOKEN_FILE);
  console.log('    token: ' + token.substring(0, 40) + '...');
  return token;
}

function loadToken() {
  if (!fs.existsSync(TOKEN_FILE)) throw new Error('未找到 token, 请先执行: node ai-api-test.js login');
  return fs.readFileSync(TOKEN_FILE, 'utf8').trim();
}

// ============ AI 接口测试 ============
async function testAi() {
  const token = loadToken();
  const headers = { Authorization: 'Bearer ' + token, clientid: CLIENT_ID };

  console.log('==> [1/4] POST /ai/chat/conversations 创建会话');
  let resp = await fetch(BASE + '/ai/chat/conversations', {
    method: 'POST', headers: { 'Content-Type': 'application/json', ...headers },
    body: JSON.stringify({ title: '接口测试会话' })
  });
  let data = await resp.json();
  console.log('    code=' + data.code + (data.msg ? ' msg=' + data.msg : ''));
  if (data.code !== 200) throw new Error('创建会话失败: ' + JSON.stringify(data));
  const conversationId = data.data.conversationId;
  console.log('    conversationId=' + conversationId);

  console.log('==> [2/4] GET /ai/chat/conversations 会话列表');
  resp = await fetch(BASE + '/ai/chat/conversations', { headers });
  data = await resp.json();
  console.log('    code=' + data.code + ', 会话数量=' + (data.data ? data.data.length : 0));

  console.log('==> [3/4] POST /ai/chat/completions 流式对话 (SSE)');
  resp = await fetch(BASE + '/ai/chat/completions', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Accept: 'text/event-stream', ...headers },
    body: JSON.stringify({ conversationId, content: '你好，请用一句话介绍你自己' })
  });
  console.log('    HTTP ' + resp.status + ', content-type=' + resp.headers.get('content-type'));
  if (!resp.ok || !resp.body) throw new Error('流式对话请求失败: HTTP ' + resp.status);
  const reader = resp.body.getReader();
  const decoder = new TextDecoder('utf-8');
  let buffer = '';
  let fullText = '';
  let done = false;
  let errMsg = null;
  while (!done) {
    const { done: streamEnd, value } = await reader.read();
    if (streamEnd) break;
    buffer += decoder.decode(value, { stream: true });
    const blocks = buffer.split('\n\n');
    buffer = blocks.pop() || '';
    for (const block of blocks) {
      if (!block.trim()) continue;
      let eventName = 'message';
      const dataLines = [];
      for (const line of block.split('\n')) {
        if (line.startsWith('event:')) eventName = line.slice(6).trim();
        else if (line.startsWith('data:')) dataLines.push(line.slice(5).replace(/^ /, ''));
      }
      const payload = dataLines.join('\n');
      if (eventName === 'error') { errMsg = payload || 'AI 服务异常'; done = true; break; }
      if (payload === '[DONE]') { done = true; break; }
      if (payload) { fullText += payload; process.stdout.write(payload); }
    }
  }
  console.log('');
  if (errMsg) {
    console.log('    [流式输出异常] ' + errMsg);
    console.log('    提示: 请检查本地模型服务 -> node ai-api-test.js ping');
  } else {
    console.log('    [流式输出完成] 共 ' + fullText.length + ' 字符');
  }

  console.log('==> [4/4] GET /ai/chat/conversations/{id}/messages 历史消息');
  resp = await fetch(BASE + '/ai/chat/conversations/' + conversationId + '/messages', { headers });
  data = await resp.json();
  console.log('    code=' + data.code + ', 消息数量=' + (data.data ? data.data.length : 0));
  if (data.data) {
    for (const msg of data.data) {
      const preview = String(msg.content || '').substring(0, 60).replace(/\n/g, ' ');
      console.log('    [' + msg.role + '] ' + preview);
    }
  }
}

// ============ 本地模型服务检查 ============
async function pingLlm() {
  console.log('==> 检查模型服务 ' + LLM_BASE + '/v1/models');
  try {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), 5000);
    const resp = await fetch(LLM_BASE + '/v1/models', { signal: controller.signal });
    clearTimeout(timer);
    const data = await resp.json();
    console.log('    HTTP ' + resp.status + ', 可用模型:');
    for (const m of data.data || []) console.log('    - ' + m.id);
    const found = (data.data || []).some((m) => m.id === LLM_MODEL);
    console.log(found ? '    [OK] 配置模型 ' + LLM_MODEL + ' 已加载' : '    [WARN] 配置模型 ' + LLM_MODEL + ' 未在列表中, 请核对 application.yml spring.ai.openai.chat.model');
  } catch (e) {
    console.log('    [FAIL] 无法连接模型服务: ' + (e.name === 'AbortError' ? '超时' : e.message));
    console.log('    提示: 请确认本地 1234 端口的模型服务(如 LM Studio)已启动并加载模型');
  }
}

// ============ 入口 ============
(async () => {
  const cmd = process.argv[2] || 'all';
  try {
    if (cmd === 'login') await login();
    else if (cmd === 'ping') await pingLlm();
    else if (cmd === 'ai') await testAi();
    else if (cmd === 'all') {
      await login();
      await pingLlm();
      await testAi();
    } else {
      console.log('未知命令: ' + cmd + ' (可选: login | ping | ai | all)');
    }
    console.log('==> 测试结束');
  } catch (e) {
    console.error('[测试失败] ' + e.message);
    process.exit(1);
  }
})();
